### PR TITLE
pdb: silence 'PodDisruptionBudgetAtLimit' alert for migratable VMIs

### PR DESCRIPTION
**Pull request by acardace**
Silence the alert only when running on OpenShift (upstream k8s does not
have this alert).

Signed-off-by: Antonio Cardace <acardace@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This patch will be effective only when https://github.com/openshift/cluster-kube-controller-manager-operator/pull/583 is merged.

**Release note**:

```release-note
silence PodDisruptionBudgetAtLimit for migratable VMIs when running on OpenShift
```

### CODE REVIEW

**Comment by acardace** [comment]
/hold
until https://github.com/openshift/cluster-kube-controller-manager-operator/pull/583 is merged

**Review by marceloamaral — COMMENTED** [review]
Why do you want to disable the alert?
Because is it annoying or we shouldn't get this alert when migrating a VMI?

Instead of hiding a potential issue, if the alert  shouldn't exist, IMHO the pod disruption budget `minAvailable`  should be updated when migrating the VMI...

In fact, virt-controller-disruption-budget adds too many keys to its workqueue, any change to decrease this rate will also improve the performance and scalability of the controller

**Comment by acardace** [comment]
> Why do you want to disable the alert? Because is it annoying or we shouldn't get this alert when migrating a VMI?

It is both annoying and confusing as an alert suggests that something is wrong and while this alert should be triggered in most situations in this for one KubeVirt is actually confusing as we're aware that we have the number of pods equal to minAvailable at all times and it is like this by design otherwise the live migration on eviction mechanism wouldn't work.

> 
> Instead of hiding a potential issue, if the alert shouldn't exist, IMHO the pod disruption budget `minAvailable` should be updated when migrating the VMI...

This is not possible the alert comes from https://github.com/openshift/cluster-kube-controller-manager-operator and while this alert is just noise for KubeVirt (for migratable VMIs PDBs) it actually makes sense in other situations as it they might prevent a cluster drain.

Also `minAvailable` is updated when migrating the VMI, take a look here https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/migration.go#L779.

> 
> In fact, virt-controller-disruption-budget adds too many keys to its workqueue, any change to decrease this rate will also improve the performance and scalability of the controller

It's not clear to me how this relates to this PR?

**Comment by fabiand** [comment]
Any timeline to get this in?

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

File: pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go
Comment (davidvossel) [inline pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go:501]: i just caught that this label is only added during creation. If you want this to occur during updates for existing VMIs, you'll need to add the label to the patch logic that expands the PDB as well.

**Review by davidvossel — COMMENTED** [review]
/approve cancel

needs update logic i think

File: pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go
Comment (acardace) [reply pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go:497]: Good catch! Didn't think about that, I've changed it now but how should we test this aspect? It would require a downgrade/upgrade e2e test, should I add that?

Comment (davidvossel) [reply pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go:497]: should be easy. Just mutate one of the pdb's in your existing functional test to remove the label and make sure it magically gets re-added

Comment (acardace) [reply pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go:497]: Done, thanks for the suggestion, can you take another look?

**Comment by rmohr** [comment]
@davidvossel @acardace do you plan to continue the work on this PR?

**Comment by acardace** [comment]
> @davidvossel @acardace do you plan to continue the work on this PR?

Hi Roman, yes that's the plan, I'm still waiting for https://github.com/openshift/cluster-kube-controller-manager-operator/pull/583 to be reviewed.

File: pkg/virt-controller/watch/application_test.go
Comment (enp0s3) [inline pkg/virt-controller/watch/application_test.go:46]: A minor one - I find the came case usage here a little bit confusing. I almost never saw us using it for aliasing. I saw either a snake case or just in small letters.

File: pkg/virt-controller/watch/migration_test.go
Comment (enp0s3) [inline pkg/virt-controller/watch/migration_test.go:152]: It looks like using the prefix approach makes the test more relaxed comparing to what it was before when it expected explicit labels.

File: tests/migration_test.go
Comment (enp0s3) [inline tests/migration_test.go:3585]: I don't think you need the `sig-compute` label since its already present in the top level spec container `Describe("[rfe_id:393]...`

File: tests/migration_test.go
Comment (enp0s3) [inline tests/migration_test.go:3592]: same here

**Review by enp0s3 — COMMENTED** [review]
@acardace Hi. Thank you! I have a few questions below. looks great! Did you had the chance to run these changes inc. the e2e test on an Openshift cluster?

File: pkg/virt-controller/watch/application_test.go
Comment (acardace) [reply pkg/virt-controller/watch/application_test.go:46]: You're right, I actually didn't notice this as it probably got added by running `make`. Will fix, thanks!

**Comment by marceloamaral** [comment]
@enp0s3 could you please fix the merge conflict of this PR?

**Comment by stu-gott** [comment]
/hold

It's not useful to run tests on this PR until merge conflicts are resolved. Also https://github.com/openshift/cluster-kube-controller-manager-operator/pull/583 is still pending.

**Pull request closed without merging**
